### PR TITLE
Natural Language: Add proper noun detection and sentence sentiment samples

### PR DESCRIPTION
**Pull request by sgreenberg**
Adjusted language_samples to demo two key features

    1 - Entities sample shows off proper noun detection.
    2 - Sentiment sample shows off sentence level sentiment.

    Fixed Tests. Addressed CR fdbk from remi.

### CODE REVIEW

File: language/language_samples.rb [hunk 1]
```diff
@@ -21,9 +21,11 @@ def sentiment_from_text project_id:, text_content:
 
   language  = Google::Cloud::Language.new project: project_id
   document  = language.document text_content
```
Comment (frankyn) [inline language/language_samples.rb]: Add a newline after this line. I know it's not part of your change, can you remove extra spaces in front of `language` and `document`.

File: language/language_samples.rb [hunk 2]
```diff
@@ -21,9 +21,11 @@ def sentiment_from_text project_id:, text_content:
...
-  sentiment = document.sentiment
-
-  puts "#{sentiment.score} (#{sentiment.magnitude})"
+  puts "Overall document sentiment: (#{document.sentiment.score})"
+  puts "Sentence level sentiment:"
```
Comment (frankyn) [inline language/language_samples.rb]: Add a new line after this line.

File: language/language_samples.rb [hunk 3]
```diff
@@ -21,9 +21,11 @@ def sentiment_from_text project_id:, text_content:
 
   language  = Google::Cloud::Language.new project: project_id
   document  = language.document text_content
-  sentiment = document.sentiment
-
-  puts "#{sentiment.score} (#{sentiment.magnitude})"
+  puts "Overall document sentiment: (#{document.sentiment.score})"
```
Comment (frankyn) [inline language/language_samples.rb]: You can reduce the number of chained method calls by creating a new variable `sentiment = document.sentiment`.

File: language/language_samples.rb [hunk 4]
```diff
@@ -21,9 +21,11 @@ def sentiment_from_text project_id:, text_content:
...
+  puts "Sentence level sentiment:"
+  document.sentiment.sentences.each do |sentence|
```
Comment (frankyn) [inline language/language_samples.rb]: Reduce this using `sentiment` variable.

File: language/language_samples.rb [hunk 5]
```diff
@@ -36,9 +38,11 @@ def sentiment_from_cloud_storage_file project_id:, storage_path:
 
   language  = Google::Cloud::Language.new project: project_id
   document  = language.document storage_path
```
Comment (frankyn) [inline language/language_samples.rb]: Add a newline after this line. I know it's not part of your change, can you remove extra spaces in front of `language` and `document`.

File: language/language_samples.rb [hunk 6]
```diff
@@ -36,9 +38,11 @@ def sentiment_from_cloud_storage_file project_id:, storage_path:
...
-  sentiment = document.sentiment
-
-  puts "#{sentiment.score} (#{sentiment.magnitude})"
+  puts "Overall document sentiment: (#{document.sentiment.score})"
```
Comment (frankyn) [inline language/language_samples.rb]: You can reduce the number of chained method calls by creating a new variable sentiment = document.sentiment.

File: language/language_samples.rb [hunk 7]
```diff
@@ -36,9 +38,11 @@ def sentiment_from_cloud_storage_file project_id:, storage_path:
...
+  puts "Sentence level sentiment:"
```
Comment (frankyn) [inline language/language_samples.rb]: Add a new line after this line.

File: language/language_samples.rb [hunk 8]
```diff
@@ -36,9 +38,11 @@ def sentiment_from_cloud_storage_file project_id:, storage_path:
...
+  document.sentiment.sentences.each do |sentence|
```
Comment (frankyn) [inline language/language_samples.rb]: Reduce `document.sentiment` to `sentiment`

File: language/language_samples.rb [hunk 9]
```diff
@@ -55,6 +59,9 @@ def entities_from_text project_id:, text_content:
 
   entities.each do |entity|
     puts "Entity #{entity.name} #{entity.type}"
```
Comment (frankyn) [inline language/language_samples.rb]: Add a new line after this line.

File: language/language_samples.rb [hunk 10]
```diff
@@ -72,6 +79,9 @@ def entities_from_cloud_storage_file project_id:, storage_path:
...
```
Comment (frankyn) [inline language/language_samples.rb]: Add a new line after this line.

File: language/language_samples.rb [hunk 11]
```diff
@@ -72,6 +79,9 @@ def entities_from_cloud_storage_file project_id:, storage_path:
...
+    if entity.metadata["wikipedia_url"]
+      puts "URL: #{entity.metadata['wikipedia_url']}"
```
Comment (frankyn) [inline language/language_samples.rb]: Tests need to be added for this metadata.

File: language/language_samples.rb [hunk 12]
```diff
@@ -55,6 +59,9 @@ def entities_from_text project_id:, text_content:
...
```
Comment (frankyn) [inline language/language_samples.rb]: Tests need to be added for this metadata.

File: language/language_samples.rb [hunk 13]
```diff
@@ -36,9 +38,11 @@ def sentiment_from_cloud_storage_file project_id:, storage_path:
 
   language  = Google::Cloud::Language.new project: project_id
   document  = language.document storage_path
-  sentiment = document.sentiment
-
-  puts "#{sentiment.score} (#{sentiment.magnitude})"
+  puts "Overall document sentiment: (#{document.sentiment.score})"
+  puts "Sentence level sentiment:"
+  document.sentiment.sentences.each do |sentence|
+    puts "#{sentence.text}: (#{sentence.sentiment.score})"
```
Comment (frankyn) [inline language/language_samples.rb]: Tests should be added to verify multiple sentences were identified by the sample.

File: language/language_samples.rb [hunk 14]
```diff
@@ -21,9 +21,11 @@ def sentiment_from_text project_id:, text_content:
 
   language  = Google::Cloud::Language.new project: project_id
   document  = language.document text_content
-  sentiment = document.sentiment
-
-  puts "#{sentiment.score} (#{sentiment.magnitude})"
+  puts "Overall document sentiment: (#{document.sentiment.score})"
+  puts "Sentence level sentiment:"
+  document.sentiment.sentences.each do |sentence|
+    puts "#{sentence.text}: (#{sentence.sentiment.score})"
```
Comment (frankyn) [inline language/language_samples.rb]: Tests should be added to verify multiple sentences were identified by the sample.

File: language/language_samples.rb (hunk 1 above)
Comment (sgreenberg) [inline language/language_samples.rb]: done

File: language/language_samples.rb (hunk 3 above)
Comment (sgreenberg) [inline language/language_samples.rb]: done

File: language/language_samples.rb (hunk 2 above)
Comment (sgreenberg) [inline language/language_samples.rb]: done.

File: language/language_samples.rb (hunk 4 above)
Comment (sgreenberg) [inline language/language_samples.rb]: done

File: language/language_samples.rb (hunk 5 above)
Comment (sgreenberg) [inline language/language_samples.rb]: done

File: language/language_samples.rb (hunk 6 above)
Comment (sgreenberg) [inline language/language_samples.rb]: done

File: language/language_samples.rb (hunk 7 above)
Comment (sgreenberg) [inline language/language_samples.rb]: done

File: language/language_samples.rb (hunk 8 above)
Comment (sgreenberg) [inline language/language_samples.rb]: done

File: language/language_samples.rb (hunk 9 above)
Comment (sgreenberg) [inline language/language_samples.rb]: done

File: language/language_samples.rb (hunk 10 above)
Comment (sgreenberg) [inline language/language_samples.rb]: done

File: language/language_samples.rb (hunk 14 above)
Comment (sgreenberg) [inline language/language_samples.rb]: done

File: language/language_samples.rb (hunk 13 above)
Comment (sgreenberg) [inline language/language_samples.rb]: done

File: language/language_samples.rb (hunk 12 above)
Comment (sgreenberg) [inline language/language_samples.rb]: done

File: language/language_samples.rb (hunk 11 above)
Comment (sgreenberg) [inline language/language_samples.rb]: done

File: language/spec/language_samples_spec.rb [hunk 15]
```diff
@@ -87,7 +121,17 @@ def capture &block
     expect(output).to include "Bob PERSON"
   end
 
-  example "entities from a file stored in Google Cloud Storage" do
+  example "entities with metadata from text" do
+    output = capture {
+      entities_from_text project_id:   @project_id,
```
Comment (frankyn) [inline language/spec/language_samples_spec.rb]: Add parenthesis similar to line 152

File: language/language_samples.rb [hunk 16]
```diff
@@ -55,6 +67,10 @@ def entities_from_text project_id:, text_content:
 
   entities.each do |entity|
     puts "Entity #{entity.name} #{entity.type}"
+
+    if entity.metadata["wikipedia_url"]
+      puts "URL: #{entity.metadata['wikipedia_url']}"
```
Comment (frankyn) [inline language/language_samples.rb]: Use double quotes instead of single-quotes.

File: language/language_samples.rb [hunk 17]
```diff
@@ -72,6 +88,10 @@ def entities_from_cloud_storage_file project_id:, storage_path:
...
```
Comment (frankyn) [inline language/language_samples.rb]: Use double-quotes instead of single-quotes

File: language/spec/language_samples_spec.rb [hunk 18]
```diff
@@ -45,20 +45,54 @@ def capture &block
...
 
-  example "sentiment from text" do
+  example "document sentiment from text" do
     positive_text = "Happy love it. I am glad, pleased, and delighted."
     negative_text = "I hate it. I am mad, annoyed, and irritated."
 
     expect {
       sentiment_from_text project_id: @project_id, text_content: positive_text
-    }.to output(/^\d\.\d+ \(\d\.\d+\)/).to_stdout
+    }.to output(/Overall document sentiment: \(\d\.\d+\)$/).to_stdout
```
Comment (frankyn) [inline language/spec/language_samples_spec.rb]: This test should expect output to include information understood from `text_content`

File: language/spec/language_samples_spec.rb [hunk 19]
```diff
@@ -45,20 +45,54 @@ def capture &block
...
 
     expect {
       sentiment_from_text project_id: @project_id, text_content: negative_text
-    }.to output(/^-\d\.\d+ \(\d\.\d+\)/).to_stdout
+    }.to output(/Overall document sentiment: \(-\d\.\d+\)$/).to_stdout
```
Comment (frankyn) [inline language/spec/language_samples_spec.rb]: This test should expect output to include information understood from `text_content`

File: language/spec/language_samples_spec.rb [hunk 20]
```diff
@@ -45,20 +45,54 @@ def capture &block
...
+
+  end
+
+  example "sentence-level sentiment from text" do
+    positive_text = "Happy love it. I am glad, pleased, and delighted."
+    negative_text = "I hate it. I am mad, annoyed, and irritated."
+
+    expect {
+      sentiment_from_text project_id: @project_id, text_content: positive_text
+    }.to output(/Sentence .*\n^.*\(\d\.\d+\)$\n^.*\(\d\.\d+\)$/).to_stdout
```
Comment (frankyn) [inline language/spec/language_samples_spec.rb]: This test should expect output to include information understood from `text_content`

File: language/spec/language_samples_spec.rb [hunk 21]
```diff
@@ -45,20 +45,54 @@ def capture &block
...
+
+    expect {
+      sentiment_from_text project_id: @project_id, text_content: negative_text
+    }.to output(/Sentence .*\n^.*\(-\d\.\d+\)$\n^.*\(-\d\.\d+\)$/).to_stdout
```
Comment (frankyn) [inline language/spec/language_samples_spec.rb]: This test should expect output to include information understood from `text_content`

File: language/spec/language_samples_spec.rb [hunk 22]
```diff
@@ -45,20 +45,54 @@ def capture &block
...
+  example "document sentiment from a file in Google Cloud Storage" do
+    upload "positive.txt", "Happy love it. I am glad, pleased, and delighted."
+    upload "negative.txt", "I hate it. I am mad, annoyed, and irritated."
+
+    expect {
+      sentiment_from_cloud_storage_file(
+        project_id:   @project_id,
+        storage_path: "gs://#{@bucket_name}/positive.txt"
+      )
+    }.to output(/Overall document sentiment: \(\d\.\d+\)$/).to_stdout
```
Comment (frankyn) [inline language/spec/language_samples_spec.rb]: This test should expect output to include information understood from `text_content`

File: language/spec/language_samples_spec.rb [hunk 23]
```diff
@@ -45,20 +45,54 @@ def capture &block
...
+
+    expect {
+      sentiment_from_cloud_storage_file(
+        project_id:   @project_id,
+        storage_path: "gs://#{@bucket_name}/negative.txt"
+      )
+    }.to output(/Overall document sentiment: \(-\d\.\d+\)$/).to_stdout
```
Comment (frankyn) [inline language/spec/language_samples_spec.rb]: This test should expect output to include information understood from `text_content`

File: language/spec/language_samples_spec.rb [hunk 24]
```diff
@@ -67,14 +101,14 @@ def capture &block
         project_id:   @project_id,
         storage_path: "gs://#{@bucket_name}/positive.txt"
       )
-    }.to output(/^\d\.\d+ \(\d\.\d+\)/).to_stdout
+    }.to output(/Sentence .*\n^.*\(\d\.\d+\)$\n^.*\(\d\.\d+\)$/).to_stdout
```
Comment (frankyn) [inline language/spec/language_samples_spec.rb]: This test should expect output to include information understood from `text_content`

File: language/spec/language_samples_spec.rb [hunk 25]
```diff
@@ -67,14 +101,14 @@ def capture &block
...
 
     expect {
       sentiment_from_cloud_storage_file(
         project_id:   @project_id,
         storage_path: "gs://#{@bucket_name}/negative.txt"
       )
-    }.to output(/^-\d\.\d+ \(\d\.\d+\)/).to_stdout
+    }.to output(/Sentence .*\n^.*\(-\d\.\d+\)$\n^.*\(-\d\.\d+\)$/).to_stdout
```
Comment (frankyn) [inline language/spec/language_samples_spec.rb]: This test should expect output to include information understood from `text_content`

File: language/language_samples.rb (hunk 16 above)
Comment (remi) [inline language/language_samples.rb]: Actually, within double-quotes, single-quotes are correct.

Keep single.

File: language/language_samples.rb (hunk 17 above)
Comment (remi) [inline language/language_samples.rb]: Actually, within double-quotes, single-quotes are correct.

Keep single.

**Comment by remi** [comment]
This looks great!  Thanks so much for the updates @sgreenberg!

When variables are aligned and `document.sentiment` --> `sentiment` where noted, this will look good to me 😄 

Thanks again!

File: language/language_samples.rb (hunk 16 above)
Comment (frankyn) [inline language/language_samples.rb]: Thanks for the feedback :)

File: language/spec/language_samples_spec.rb (hunk 18 above)
Comment (remi) [inline language/spec/language_samples_spec.rb]: @sgreenberg It was hard to see the difference between `\(\d\.\d+\)` and `\(-\d\.\d+\)` so it caused a little confusion

Vote: document the difference so it's easier to understand.  Inotherwords, go back to a well-named local variable, just like you had in the first place 😄 

Thought:

```ruby
positive_text   = "Happy love it. I am glad, pleased, and delighted."
negative_text   = "I hate it. I am mad, annoyed, and irritated."
positive_output = /\(\d\.\d+\)/
negative_output = /\(-\d\.\d+\)/

expect {
  # ...
```

File: language/language_samples.rb [hunk 26]
```diff
@@ -19,11 +19,17 @@ def sentiment_from_text project_id:, text_content:
 
   require "google/cloud/language"
 
-  language  = Google::Cloud::Language.new project: project_id
-  document  = language.document text_content
+  language = Google::Cloud::Language.new project: project_id
```
Comment (sgreenberg) [inline language/language_samples.rb]: done

File: language/language_samples.rb [hunk 27]
```diff
@@ -19,11 +19,17 @@ def sentiment_from_text project_id:, text_content:
...
+  document = language.document text_content
```
Comment (sgreenberg) [inline language/language_samples.rb]: done

File: language/language_samples.rb [hunk 28]
```diff
@@ -19,11 +19,17 @@ def sentiment_from_text project_id:, text_content:
...
   sentiment = document.sentiment
 
-  puts "#{sentiment.score} (#{sentiment.magnitude})"
+  puts "Overall document sentiment: (#{sentiment.score})"
+  puts "Sentence level sentiment:"
+
+  document.sentiment.sentences.each do |sentence|
```
Comment (sgreenberg) [inline language/language_samples.rb]: done

File: language/language_samples.rb [hunk 29]
```diff
@@ -34,11 +40,17 @@ def sentiment_from_cloud_storage_file project_id:, storage_path:
 
   require "google/cloud/language"
 
-  language  = Google::Cloud::Language.new project: project_id
-  document  = language.document storage_path
+  language = Google::Cloud::Language.new project: project_id
```
Comment (sgreenberg) [inline language/language_samples.rb]: done

File: language/language_samples.rb [hunk 30]
```diff
@@ -34,11 +40,17 @@ def sentiment_from_cloud_storage_file project_id:, storage_path:
...
+  document = language.document storage_path
```
Comment (sgreenberg) [inline language/language_samples.rb]: done

File: language/language_samples.rb [hunk 31]
```diff
@@ -34,11 +40,17 @@ def sentiment_from_cloud_storage_file project_id:, storage_path:
...
   sentiment = document.sentiment
 
-  puts "#{sentiment.score} (#{sentiment.magnitude})"
+  puts "Overall document sentiment: (#{sentiment.score})"
+  puts "Sentence level sentiment:"
+
+  document.sentiment.sentences.each do |sentence|
```
Comment (sgreenberg) [inline language/language_samples.rb]: done

File: language/language_samples.rb (hunk 16 above)
Comment (sgreenberg) [inline language/language_samples.rb]: Cool!

File: language/language_samples.rb (hunk 17 above)
Comment (sgreenberg) [inline language/language_samples.rb]: Cool!

**Comment by frankyn** [comment]
Thanks @sgreenberg! :)

**Pull request merged**